cpp/src/io/orc/writer_impl.cu

```diff
@@ -1073,23 +1073,49 @@ void set_stat_desc_leaf_cols(device_span<orc_column_device_view const> columns,
                    [=] __device__(auto idx) { stat_desc[idx].leaf_column = &columns[idx]; });
 }
 
-writer::impl::encoded_statistics writer::impl::gather_statistic_blobs(
-  statistics_freq stats_freq,
+hostdevice_vector<uint8_t> allocate_and_encode_blobs(
+  hostdevice_vector<statistics_merge_group>& stats_merge_groups,
+  rmm::device_uvector<statistics_chunk>& stat_chunks,
+  int num_stat_blobs,
+  rmm::cuda_stream_view stream)
+{
+  // figure out the buffer size needed for protobuf format
+  gpu::orc_init_statistics_buffersize(
+    stats_merge_groups.device_ptr(), stat_chunks.data(), num_stat_blobs, stream);
+  stats_merge_groups.device_to_host(stream, true);
+
+  hostdevice_vector<uint8_t> blobs(stats_merge_groups[num_stat_blobs - 1].start_chunk +
+                                     stats_merge_groups[num_stat_blobs - 1].num_chunks,
+                                   stream);
+  gpu::orc_encode_statistics(blobs.device_ptr(),
+                             stats_merge_groups.device_ptr(),
+                             stat_chunks.data(),
+                             num_stat_blobs,
+                             stream);
+  stats_merge_groups.device_to_host(stream);
```

Comment:
Silly question: Why is this step (i.e. copying the merge groups back to host necessary) necessary?

Reply:
Ah, `orc_encode_statistics()` does modify the merge-groups.
I think I'm still missing why this isn't `device_to_host(stream, true)`.

Reply:
There is a pile of back and forth from device in these functions. :(

Reply:
> I think I'm still missing why this isn't device_to_host(stream, true).

The sync comes after the next line:
`blobs.device_to_host(stream, true);`

We don't need to sync after each, but need to ensure everything is transferred at the bottom of the function.